### PR TITLE
Fixes #1400, #612, #1632, #1634, #1531, #1458: Tab isn't handled properly for insert and visualblockinsert modes

### LINKED ISSUE #612: Repeated commands that insert white space at from of line doesn't work
Run: `I\t\t<esc>...`

It successfully inserts tabs in the first part (the `I\t\t<esc>` block), but repeating the command with `...` does not successfully append 2 tabs to the front of the line.

### LINKED ISSUE #1400: Tab in visual block mode applies to first line only
Please *thumbs-up* 👍 this issue if it personally affects you! You can do this by clicking on the emoji-face on the top right of this post. Issues with more thumbs-up will be prioritized.

-----

### What did you do?

In a document with multiple lines and the cursor at the beginning of the first line;
`Ctrl+v`, cursor down to include all lines, `Shift+i`, `Tab`, `Esc`

### What did you expect to happen?
All lines that was included in the visual block has white space added

### What happened instead?
Only the first line had white space added.
It works if I use `Space` instead of `Tab`.

### Technical details:

* VSCode Version: 1.10.2
* VsCodeVim Version: 0.6.6
* OS: Win 7

### LINKED ISSUE #1458: Inserting keywords on pressing arrow keys in visual mode
### What did you do?

**Ctrl+V** (enter visual mode), select a few lines, type **I**, type # (just to comment these lines), press Up / Down arrow (for example)

### What did you expect to happen?

Just go to a line above / below

### What happened instead?

On selected lines appears _<up>_ and _<down>_

### Technical details:

* VSCode Version: 1.10.2
* VsCodeVim Version: 0.6.11
* OS: Windown 7 x64

### LINKED ISSUE #1634: macro playback does not insert tab
* Click *thumbs-up* 👍 on this issue if you want it!
* Click *confused* 😕 on this issue if not having it makes VSCodeVim unusable. 

The VSCodeVim team prioritizes issues based on reaction count. 

--------

**BUG REPORT**:

**Environment**:

- **VSCode Version: 1.11.2**:
- **VsCodeVim Version: 0.6.20**: 
- **OS: Windows 10 Home**: 

**What happened**:
Tab is not inserted on macro replay. Record a macro that goes to beginning of a line, enters insert mode, inserts one tab, exits insert mode, moves down one line. Replay the macro.

`qa0i<tab><esc>jq@a`

**What did you expect to happen**: 
1. Cursor moves to beginning of line
2. Enters insert mode
3. Inserts one tab
4. Exits insert mode
5. Moves down one line

**How to reproduce it**:
Record macro:
`qa0i<tab><esc>jq`

Replay macro:
`@a`

### PR DESCRIPTION
**Pull request by Chillee**
So really, what I was doing in these commits was refactoring Visual Block Mode to use multicursors instead, where I also added my own code for handling Insert Mode Tab.

http://vimdoc.sourceforge.net/htmldoc/visual.html#blockwise-operators

Thus, this PR fixes (some of) #776, fixes #1400, fixes #612, fixes #1458, and fixes #1634.

I also added fixing #1531 into this one since its editing is a similar concept to visualblock.

### CODE REVIEW

**Comment by johnfn** [comment]
Wow, this is great! Is it ready to go?

**Comment by Chillee** [comment]
@johnfn Should be. I just realized that this PR actually also contains this PR: https://github.com/VSCodeVim/Vim/pull/1659, but that PR should also be fairly uncontroversial.

Lemme add a couple of tests for the new o/O functionality, and then I'd be fine with this being merged.

**Comment by Chillee** [comment]
@johnfn Alright just added tests. It should be ready to merge.

**Comment by Chillee** [comment]
This is a very bizarre bug.

This is the test that's failing:
```
|foo
```
with the keys `yyO<esc>p`.
This results in 
```

foo
|foo
```
(this is because of the cursor position; it's not skipping a line when it pastes. For example, if bar was in the register, it would be
```

bar
|foo
```
).

The bizarre part about this is that if the cursor is like this before the `O`:
```
f|oo
```
it works fine. Looking at the returned vimState between the two cases, they look identical.

**Comment by Chillee** [comment]
I feel like it *has* to be an upstream issue, with how weird this bug is.

**Comment by Chillee** [comment]
I ended up fixing it by moving the cursor to the right before calling `editor.action.insertLineBefore`

File: src/actions/commands/actions.ts [hunk 1]
```diff
@@ -462,7 +456,7 @@ class CommandEsc extends BaseCommand {
     }
 
     if (vimState.currentMode === ModeName.EasyMotionMode) {
-      // Escape or other termination keys were pressed, exit mode
+      // Escaape or other termination keys were pressed, exit mode
```
Comment (jpoon) [inline src/actions/commands/actions.ts]: typo

**Comment by johnfn** [comment]
Is this good to go now? 😁

**Comment by Chillee** [comment]
Except for that typo...

**Comment by johnfn** [comment]
The cool thing is that as a contributor I can fix that directly from GitHub. Here we gooo

**Comment by Chillee** [comment]
The reason I didn't say I completely fixed that refactor issue is that there's some visual block code I still haven't touched, related to visual block mode not in insert.

**Pull request merged**